### PR TITLE
[Testing] Run slowest tests early

### PR DESCRIPTION
**Pull request by davezarzycki**

### CODE REVIEW

**Comment by shahmishal** [comment]
https://github.com/apple/llvm-project/pull/2478
@swift-ci test

File: test/lit.cfg [hunk 1]
```diff
@@ -322,6 +322,30 @@ config.round_trip_syntax_test = make_path(config.swift_utils, 'round-trip-syntax
 config.link = lit.util.which('link', config.environment.get('PATH', '')) or      \
               lit.util.which('lld-link', config.environment.get('PATH', ''))
 
+config.slowest_tests = {
```
Comment (benrimmington) [inline test/lit.cfg:325]:
```suggestion
config.early_tests = {
```

Comment (davezarzycki) [reply test/lit.cfg:325]: Thanks. I'm still actually discussing a refinement to upstream (which converged on `early_tests`). I had wanted to add `late_tests` to let us schedule the "complier crasher" tests late, but now I'm of the opinion that we should just have `test_phases` that is flexible enough to allow minimal maintenance for those that want it. If you want to follow along: https://reviews.llvm.org/D97046

**Comment by davezarzycki** [comment]
So good news. After some back and forth with the LLVM community, we settled on a way better solution than manually managed early/late test lists. Now `lit` can just learn what is slow and and what is fast from prior testing runs and then reorder tests appropriately for the next test run. This doesn't automatically help bots but one can commit some or all of the timing data back to the repo if that matters.

I'll leave this open for now, in case we want to commit some timing data after the following cherry-pick is merged:

https://github.com/apple/llvm-project/pull/2687

**Comment by benrimmington** [comment]
> This doesn't automatically help bots but one can commit some or all of the timing data back to the repo if that matters.

When a _benchmark_ is created or updated, how is the new baseline shared with all the bots?
Could the same mechanism be used to share the `.lit_test_times.txt` file?

**Comment by davezarzycki** [comment]
Can you give an example? And/or rephrase the question? Benchmarking is a surprisingly complicated topic.

**Comment by benrimmington** [comment]
I'm sorry for not being clear. I was referring to the [Swift benchmark suite](https://github.com/apple/swift/tree/main/benchmark).

* In a pull request to add a new benchmark, running `@swift-ci Please benchmark` will show the results under an "Added" heading (rather than a "Regression" or "Improvement" heading). When the pull request is merged, the baseline results for that new benchmark are _somehow_ shared with all the bots.

* In a separate pull request to modify the standard library, which may affect the performance of that recently added benchmark, running `@swift-ci Please benchmark` will now show the results relative to the updated baseline (i.e. an improvement or regression).

The benchmark baseline is a text file, stored in the build directory, so it seems similar to your `.lit_test_times.txt` file.

**Comment by davezarzycki** [comment]
Okay. I'm not sure how the Swift CI benchmarking works. If CI is creating a really quiet and consistent environment for benchmarks to run, then the "record and reorder" `lit` change mentioned earlier shouldn't be observable. Also it really depends on how clean the build environment from CI job to CI job. (I'd like to point out that CI supports "clean test", which implies some amount of non-cleanliness is the default behavior.)

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Comment by shahmishal** [comment]
This will fail because automerger has not merged the changes into `swift/main`, I will restart it once testing once it's merged.

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Comment by davezarzycki** [comment]
@shahmishal – The patch is stale (but GitHub doesn't realize that). Now that `lit` can automatically reorder tests, this patch isn't needed for people's individual workflows so we could just abandon this patch. On the other hand, if you want, we could gather timing data and commit some of it in an attempt to make the bots faster. What do you think?

**Comment by shahmishal** [comment]
Sounds good, we can close this PR.

**Pull request closed without merging**